File: scripts/merge-qol-packages.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import os
from pathlib import Path, PurePosixPath
import sys
from zipfile import BadZipFile, ZIP_DEFLATED, ZipFile, ZipInfo


REQUIRED_NATIVE_LIBRARIES = (
    "Code/lib-win-x64/microblocks_qol_native.dll",
    "Code/lib-linux/libmicroblocks_qol_native.so",
    "Code/lib-osx/libmicroblocks_qol_native.dylib",
)
CANONICAL_BASE_FILES = {"Code/MicroblocksQolUtils.dll"}


def archive_path(name: str) -> str:
    normalized = name.replace("\\", "/").lstrip("/")
    path = PurePosixPath(normalized)
    if not normalized or path.is_absolute() or ".." in path.parts:
        raise ValueError(f"unsafe archive entry: {name}")
    return path.as_posix()


def merge_packages(output: Path, packages: list[Path]) -> None:
    files: dict[str, bytes] = {}
    origins: dict[str, Path] = {}

    for package in packages:
        with ZipFile(package, "r") as archive:
            for entry in archive.infolist():
                if entry.is_dir():
                    continue
                name = archive_path(entry.filename)
                contents = archive.read(entry)
                if name in files and files[name] != contents:
                    if name in CANONICAL_BASE_FILES:
                        continue
                    raise ValueError(
                        f"platform packages contain different shared file contents: {name} "
                        f"({origins[name].name} and {package.name})"
                    )
                files[name] = contents
                origins.setdefault(name, package)

    missing = [name for name in REQUIRED_NATIVE_LIBRARIES if name not in files]
    if missing:
        raise ValueError(f"merged package is missing: {', '.join(missing)}")

    output.parent.mkdir(parents=True, exist_ok=True)
    temporary = output.with_name(output.name + ".tmp")
    temporary.unlink(missing_ok=True)
    try:
        with ZipFile(temporary, "w", compression=ZIP_DEFLATED, compresslevel=9) as archive:
            for name in sorted(files):
                entry = ZipInfo(name, date_time=(1980, 1, 1, 0, 0, 0))
                entry.compress_type = ZIP_DEFLATED
                entry.external_attr = 0o100644 << 16
                archive.writestr(entry, files[name], compresslevel=9)
        os.replace(temporary, output)
    finally:
        temporary.unlink(missing_ok=True)

    print(f"Merged {len(packages)} platform packages into {output}")


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Merge platform-specific MicroblocksQolUtils packages into one archive."
    )
    parser.add_argument("output", type=Path)
    parser.add_argument("packages", nargs="+", type=Path)
    arguments = parser.parse_args()

    output = arguments.output.resolve()
    packages = [package.resolve() for package in arguments.packages]
    if output in packages:
        parser.error("the output archive cannot also be an input archive")
    for package in packages:
        if not package.is_file():
            parser.error(f"input archive does not exist: {package}")

    try:
        merge_packages(output, packages)
    except (BadZipFile, OSError, ValueError) as error:
        print(error, file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/merge-qol-packages.mjs
import {
  existsSync,
  mkdirSync,
  mkdtempSync,
  readdirSync,
  readFileSync,
  rmSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative, resolve, sep } from "node:path";
import { tmpdir } from "node:os";
import { spawnSync } from "node:child_process";

const [outputArgument, ...packageArguments] = process.argv.slice(2);
if (!outputArgument || packageArguments.length === 0) {
  console.error("Usage: node scripts/merge-qol-packages.mjs OUTPUT.zip PLATFORM-PACKAGE.zip...");
  process.exit(2);
}

const output = resolve(outputArgument);
const packages = packageArguments.map((path) => resolve(path));
if (packages.includes(output)) throw new Error("The output archive cannot also be an input archive");
for (const packagePath of packages) {
  if (!existsSync(packagePath) || !statSync(packagePath).isFile()) {
    throw new Error(`Input archive does not exist: ${packagePath}`);
  }
}

const run = (command, args, options = {}) => {
  const result = spawnSync(command, args, { stdio: "inherit", ...options });
  if (result.error) throw result.error;
  if (result.status !== 0) throw new Error(`${command} failed with status ${result.status}`);
};

const files = new Map();
const canonicalBaseFiles = new Set(["Code/MicroblocksQolUtils.dll"]);
const work = mkdtempSync(join(tmpdir(), "microblocks-qol-merge-"));

const walkFiles = (root) => {
  const result = [];
  const visit = (directory) => {
    for (const entry of readdirSync(directory, { withFileTypes: true })) {
      const path = join(directory, entry.name);
      if (entry.isDirectory()) visit(path);
      else if (entry.isFile()) result.push(path);
      else throw new Error(`Unsupported archive entry type after extraction: ${path}`);
    }
  };
  visit(root);
  return result;
};

try {
  for (const [index, packagePath] of packages.entries()) {
    const extracted = join(work, `package-${index}`);
    mkdirSync(extracted);
    run("unzip", ["-q", packagePath, "-d", extracted]);
    for (const source of walkFiles(extracted)) {
      const name = relative(extracted, source).split(sep).join("/");
      if (!name || name.startsWith("/") || name.split("/").includes("..")) {
        throw new Error(`Unsafe archive entry after extraction: ${name}`);
      }
      const contents = readFileSync(source);
      const previous = files.get(name);
      if (previous && !previous.contents.equals(contents)) {
        if (canonicalBaseFiles.has(name)) continue;
        throw new Error(`Platform packages contain different shared file contents: ${name}`);
      }
      files.set(name, { contents });
    }
  }

  for (const required of [
    "Code/lib-win-x64/microblocks_qol_native.dll",
    "Code/lib-linux/libmicroblocks_qol_native.so",
    "Code/lib-osx/libmicroblocks_qol_native.dylib",
  ]) {
    if (!files.has(required)) throw new Error(`Merged package is missing ${required}`);
  }

  const merged = join(work, "merged");
  for (const [name, { contents }] of [...files.entries()].sort(([a], [b]) => a.localeCompare(b))) {
    const destination = join(merged, ...name.split("/"));
    mkdirSync(dirname(destination), { recursive: true });
    writeFileSync(destination, contents);
  }

  mkdirSync(dirname(output), { recursive: true });
  rmSync(output, { force: true });
  const entries = readdirSync(merged).sort();
  run("zip", ["-q", "-r", output, ...entries], { cwd: merged });
  console.log(`Merged ${packages.length} platform packages into ${output}`);
} finally {
  rmSync(work, { recursive: true, force: true });
}
